### PR TITLE
Unit test for deployments API

### PR DESCRIPTION
**Pull request by supritat**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### CODE REVIEW

File: jazz_deployments/index.js [hunk 1]
```diff
@@ -159,15 +159,13 @@ module.exports.handler = (event, context, cb) => {
...
-			} else if (error.result === "unauthorized") {
-				return cb(JSON.stringify(errorHandler.throwUnauthorizedError(error.message)));
 			} else {
 				return cb(JSON.stringify(errorHandler.throwInternalServerError("Unexpected error occurred.")))
 			}
 		});
 };
 
-function genericInputValidation(event) {
+var genericInputValidation = (event) => {
```
Comment (svidhani) [inline jazz_deployments/index.js]: why did you change all these to be function expression. you could have just exported functions

File: jazz_deployments/test/test.js [hunk 2]
```diff
@@ -1,11 +1,1106 @@
...
+        }
+      }
+    };
+    AWS.mock("DynamoDB.DocumentClient", "query", (param, cb) => {
+      var dataObj = {
+        Items: [event.body]
+      }
+      return cb(null, dataObj);
+    });
+    reqStub = sinon.stub(request, "Request", (obj) => {
```
Comment (svidhani) [inline jazz_deployments/test/test.js]: please also verify that the stub was being called eg: sinon.assert.ccalledOnce(reqStub).
please update in other places as well where you are using stubs or spies

File: jazz_deployments/test/test.js [hunk 3]
```diff
@@ -1,11 +1,1106 @@
...
+const request = require('request');
+const sinon = require('sinon');
+
 const index = require('../index');
+const logger = require("../components/logger.js");
+const configObj = require('../components/config.js');
+const crud = require('../components/crud')();
+const errorHandler = require("../components/error-handler.js")();
+
+describe('jazz_deployments', function () {
```
Comment (svidhani) [inline jazz_deployments/test/test.js]: you can have multiple "describes" one for each function so its easy to group those test suites to track them and update them. Having them in one describe is hard to identify what tests are testing which unit.

Comment (supritat) [reply jazz_deployments/test/test.js]: Worked as per comments.

File: jazz_deployments/test/test.js (hunk 2 above)
Comment (supritat) [reply jazz_deployments/test/test.js]: Worked as per comments

File: jazz_deployments/test/test.js [hunk 4]
```diff
@@ -1,11 +1,1181 @@
...
+      event.method = "POST";
+      event.path =  {};
+      message = '{"errorType":"InternalServerError","message":"unexpected error occurred"}';
+      AWS.mock("DynamoDB.DocumentClient", "put", (params, cb) => {
+        return cb(err, null);
+      });
+      index.handler(event, context,(err, res) => {
+        err.should.be.equal(message);
+        AWS.restore("DynamoDB.DocumentClient");
+        return res;
```
Comment (svidhani) [inline jazz_deployments/test/test.js]: why do you have to return here? remove return res from all the test where it is not used to assert later

File: jazz_deployments/test/test.js [hunk 5]
```diff
@@ -1,11 +1,1181 @@
...
+        }
+      };
+      AWS.mock("DynamoDB.DocumentClient", "query", (param, cb) => {
+        var dataObj = {
+          Items: [event.body]
+        }
+        return cb(null, dataObj);
+      });
+      reqStub = sinon.stub(request, "Request", (obj) => {
+        console.log(obj);
```
Comment (svidhani) [inline jazz_deployments/test/test.js]: remove console everywhere

File: jazz_deployments/test/test.js [hunk 6]
```diff
@@ -1,11 +1,1181 @@
...
+        if(obj.method.toLowerCase() === 'post') {
```
Comment (svidhani) [inline jazz_deployments/test/test.js]: I would split this into two tests and not do if/else. set it with post in one and get in another and make it predictable rather than using if/else.

File: jazz_deployments/test/test.js [hunk 7]
```diff
@@ -1,11 +1,1181 @@
...
+        var dataObj = {
+          Items: [event.body]
+        }
+        return cb(null, dataObj);
+      });
+      var i=0;
+      reqStub = sinon.stub(request, "Request", (obj) => {
+        i++;
+        console.log(i)
+        if(obj.method.toLowerCase() === 'post') {
```
Comment (svidhani) [inline jazz_deployments/test/test.js]: same comment here. please don't do if/else. let the test do one thing only. please update in other tests as well

File: jazz_deployments/test/test.js [hunk 8]
```diff
@@ -1,11 +1,1181 @@
...
+      });
+      index.handler(event, context, (err, res) => {
+        err.should.be.equal(message);
+        AWS.restore("DynamoDB");
+        return err;
+      })
+    });
+  
+    it("should successfully get deployment with provided path param using GET method", () => {
+      event.method = "GET";
```
Comment (svidhani) [inline jazz_deployments/test/test.js]: if you want you can add more describes here and make it simpler.
One each for post, get and delete. I see the setups are being repeated for each one of them and it can be done once(using beforeEach and afterEach) and group all these HTTP verbs in their own respective describes. What do you think?

File: jazz_deployments/index.js (hunk 1 above)
Comment (supritat) [reply jazz_deployments/index.js]: Worked as per comments

File: jazz_deployments/test/test.js [hunk 9]
```diff
@@ -1,11 +1,1181 @@
...
+        event.body.status = update_data.status;
+        return cb(null, dataObj);
+      });
+      AWS.mock("DynamoDB.DocumentClient", "update", (params, cb) => {
+        var dataObj = {
+          Items: [event.body]
+        };
+        return cb(null, dataObj);
+      });
+      var processDeploymentsUpdate = index.processDeploymentsUpdate(config, event.body, tableName, event.path.id);
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_deployments/test/test.js]: Worked as per comments

File: jazz_deployments/test/test.js (hunk 4 above)
Comment (supritat) [reply jazz_deployments/test/test.js]: Removed

File: jazz_deployments/test/test.js (hunk 5 above)
Comment (supritat) [reply jazz_deployments/test/test.js]: Removed console

File: jazz_deployments/test/test.js (hunk 6 above)
Comment (supritat) [reply jazz_deployments/test/test.js]: Worked as per comments

File: jazz_deployments/test/test.js (hunk 7 above)
Comment (supritat) [reply jazz_deployments/test/test.js]: Worked as per comments

File: jazz_deployments/test/test.js (hunk 8 above)
Comment (supritat) [reply jazz_deployments/test/test.js]: Worked as mentioned

File: jazz_deployments/index.js [hunk 10]
```diff
@@ -544,4 +548,28 @@ function buildNowRequest(serviceDetails, config, refDeployment) {
 			reject("unable to find deployment details");
 		}
 	});
-}
\ No newline at end of file
+}
+
+const factory = {
```
Comment (suryajak) [inline jazz_deployments/index.js]: This variable name is confusing as it isn't a `factory` in traditional terms.

File: jazz_deployments/index.js [hunk 11]
```diff
@@ -544,4 +548,28 @@ function buildNowRequest(serviceDetails, config, refDeployment) {
...
+	validateUpdateInput,
+	updateDeploymentDetails,
+	deleteServiceByID,
+	reBuildDeployment,
+	getToken,
+	getServiceDetails,
+	buildNowRequest
+}
+
+module.exports = factory;
```
Comment (suryajak) [inline jazz_deployments/index.js]: Looking at the rest of the code, looks like we can do away with this confusing variable `factory`. Can you remove it and just have that object be directly assigned here?

File: jazz_deployments/index.js [hunk 12]
```diff
@@ -30,7 +30,7 @@ const crud = require("./components/crud")(); //Import the crud module.
 const request = require('request');
 const util = require('util');
 
-module.exports.handler = (event, context, cb) => {
+function handler (event, context, cb) {
```
Comment (suryajak) [inline jazz_deployments/index.js]: Can we just use the standard convention we have in the rest of our code?
`var x = function() { ...`

Comment (svidhani) [reply jazz_deployments/index.js]: This does not work when you have to stub it.

File: jazz_deployments/index.js (hunk 11 above)
Comment (svidhani) [reply jazz_deployments/index.js]: You cannot return the functions directly in module.exports as that cannot be stubbed using sinon(in same module). The way to work around this is to pass an object that returns the functions. If naming is the issue we can come up with another name.

File: jazz_deployments/index.js [hunk 13]
```diff
@@ -217,10 +214,10 @@ function genericInputValidation(event) {
 	});
 };
 
-function processDeploymentCreation(config, deployment_details, deploymentTableName) {
+function processDeploymentCreation (config, deployment_details, deploymentTableName) {
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_deployments/index.js]: Sure

File: jazz_deployments/index.js (hunk 11 above)
Comment (supritat) [reply jazz_deployments/index.js]: Variable name is replaced as mentioned. Second part will be addressed as part of another PR.

File: jazz_deployments/index.js (hunk 10 above)
Comment (supritat) [reply jazz_deployments/index.js]: Done

File: jazz_deployments/test/test.js [hunk 14]
```diff
@@ -1,11 +1,1275 @@
...
+    it("should indicate missing required data error while validating create payload for new deployment", () => {
+      event.query = {};
+      event.path = {};
+      event.body.service = "";
+      index.validateDeploymentDetails(config, event.body)
+        .catch(error => {
+          expect(error).to.include({
+            result: 'inputError'
+          });
+        });
```
[the remark this answers is not in the document]
Comment (suryajak) [reply jazz_deployments/test/test.js]: Similar test cases are missing for _list_ and _update_ scenarios

File: jazz_deployments/test/test.js [hunk 15]
```diff
@@ -1,11 +1,1275 @@
...
+    });
```
[the remark this answers is not in the document]
Comment (suryajak) [reply jazz_deployments/test/test.js]: Similar test cases are missing for _list_ and _update_ scenarios

File: jazz_deployments/test/test.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (suryajak) [reply jazz_deployments/test/test.js]: @supritat:  Did you get a chance to look at this?

[the remark this answers is not in the document]
Comment (supritat) [reply jazz_deployments/test/test.js]: Latest merge includes the expected test cases
